Source/Core/Common/BitFieldView.h

```diff
@@ -0,0 +1,788 @@
+struct AssertHostTypeIsSane
+{
+  AssertHostTypeIsSane() = default;
+  static_assert(std::is_integral_v<host_t> || std::is_enum_v<host_t>,
+                "Given host type is not sane.  You must define a partial specialization for it.");
+};
+template <typename field_t>
+struct AssertFieldTypeIsSane
+{
+  AssertFieldTypeIsSane() = default;
+  static_assert(std::is_integral_v<field_t> || std::is_enum_v<field_t>,
+                "Given field type is not sane.  It must be an integral or enumerated type.");
+};
+template <typename field_t, std::size_t start, std::size_t width, typename host_t>
+struct AssertBitFieldFixedIsSane
+{
+  AssertBitFieldFixedIsSane() = default;
+  static_assert(start <= 8 * sizeof(host_t), "BitField out of range");
+  static_assert(start + width <= 8 * sizeof(host_t), "BitField out of range");
+  static_assert(width <= 8 * sizeof(field_t), "Field is too wide");
+  static_assert(width > 0, "BitField is invalid (zero width)");
+  static_assert(!std::is_same_v<field_t, bool> || width == 1,
+                "Boolean fields should only be 1 bit wide");
+};
+template <typename field_t, std::size_t start, std::size_t width, std::size_t length,
+          typename host_t>
+struct AssertBitFieldFixedArrayIsSane
+{
+  AssertBitFieldFixedArrayIsSane() = default;
+  static_assert(start <= 8 * sizeof(host_t), "BitFieldArray out of range");
+  static_assert(start + width * length <= 8 * sizeof(host_t), "BitFieldArray out of range");
+  static_assert(width <= 8 * sizeof(field_t), "Field is too wide");
+  static_assert(width > 0, "Field is invalid (zero width)");
+  static_assert(!std::is_same_v<field_t, bool> || width == 1,
+                "Boolean fields should only be 1 bit wide");
+};
+template <typename field_t, typename host_t>
+struct AssertBitFieldIsSane
+{
+  AssertBitFieldIsSane() = delete;
+  AssertBitFieldIsSane(const std::size_t start, const std::size_t width)
+  {
+    DEBUG_ASSERT(start <= 8 * sizeof(host_t));          // BitField out of range
+    DEBUG_ASSERT(start + width <= 8 * sizeof(host_t));  // BitField out of range
+    DEBUG_ASSERT(width <= 8 * sizeof(field_t));         // BitField is too wide
+    DEBUG_ASSERT(width > 0);                            // BitField is invalid (zero width);
+    if constexpr (std::is_same_v<field_t, bool>)
+      DEBUG_ASSERT(width == 1);  // Boolean fields should only be 1 bit wide
+  }
+};
+
+///////////////////////////////////////////////////////////////////////////////////////////////////
+// Simple, functional programming.
+
+template <typename field_t, std::size_t start, std::size_t width, typename host_t>
+constexpr field_t GetBitFieldFixed(const host_t host)
+{
+  AssertHostTypeIsSane<host_t>();
+  AssertFieldTypeIsSane<field_t>();
+  AssertBitFieldFixedIsSane<field_t, start, width, host_t>();
+
+  using uhost_t = std::make_unsigned_t<host_t>;
+  using shost_t = std::make_signed_t<host_t>;
+  constexpr std::size_t rshift = 8 * sizeof(host_t) - width;
```

Comment:
```suggestion
  constexpr std::size_t rshift = CHAR_BIT * sizeof(host_t) - width;
```

You will need `<climits>`. Ditto other instances.

Reply:
Perhaps instead consider: `Common::BitSize<host_t>()`.